src/security/CertGadgets.cc

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 1996-2019 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "Debug.h"
+#include "sbuf/SBuf.h"
+#include "security/CertGadgets.h"
+
+#if USE_OPENSSL
+#if HAVE_OPENSSL_X509V3_H
+#include <openssl/x509v3.h>
+#endif
+#endif
+
+SBuf
+Security::CertSubjectName(const Security::CertPointer &cert)
```

Comment:
Using the certificate object directly is dangerous:

* In the builds where neither library is used the CertPointer is a nil void*. Dereferencing to get the data structure of a cert is a bad idea.

* In OpenSSL builds access to the data structure is strictly reference counted. It is only the guaranteed presence of the CertPointer through the function lifetime which ensures the data structure being accessed remains present. It may otherwise be optimized to free() too early if last use of the CertPointer object in the caller was to dereference it for passing to the function).

Reply:
> Using the certificate object directly is dangerous

It is not dangerous -- in a correct C++ program, nothing can go wrong with an object passed by reference to a function during that function execution. It would be simply impossible to use references otherwise. Specifically:

> Dereferencing to get the data structure of a cert is a bad idea.

If the proposed change is implemented, there will be no dereferencing of a pointer in the adjusted function (because there will be no pointers to dereference). The caller will need to dereference, of course, but one dereferencing is unavoidable in any design and, compared to a function that must return SBuf, the caller can actually handle nil pointers _correctly_.

> It may otherwise be optimized to free() too early if last use of the CertPointer object in the caller was to dereference it for passing to the function).

C++ does not work that way. The [smart pointer] object lifetime is guaranteed to last until the end of the scope where it was declared. See "end of scope, for objects with automatic storage duration" in [Explanation](https://en.cppreference.com/w/cpp/language/destructor#Explanation). This is why various scope guards/locks work well in C++ -- the compiler cannot shorten their lifetime even though they are "unused" beyond their declaration.

There is also a related Core [Guideline](https://isocpp.github.io/CppCoreGuidelines#Rf-in).

Reply:
You are talking about the Pointer object holding a reference to the cert during the entire time the function is called yes?

I am talking about the cert handle the Pointer stores possibly being invalidated randomly by something else the library does or other applications (including Squid workers) using the same shared library and cert. This is part of why every API call to GnuTLS has an error output check.